Run smarten on PR merge

BillWagner opened the pull request:
Run the smart quoter when a PR is merged into the version release branch.

The strategy I took is to add the smarten app to the current workflow that runs when a PR is merged. That guarantees order, and ensures we have at most one automation PR to merge when a human-authored PR is merged.

Because the smarten executable is a .NET Framework based (Windows only) application, that rippled into a few other changes:

- The merge workflow now runs on Windows.
- Because the workflow runs on a Windows host, the default shell needs to specify bash.

I made the choice to run the smarten tool first of the automated transforms. That seems to be the best  choice.

BillWagner committed:
ready to test locally

BillWagner committed:
finish local script testing

BillWagner commented:
@Nigel-Ecma One interesting result from testing this locally. Both classes.md and expressions.md wouldn't round trip correctly, so a second file was created instead of overwriting the original. When I looked at the diffs, none of the quotes changed in either file. Is that consistent with your experience? If not, any ideas on how to fix?

jskeet commented on tools/run-smarten.sh:
```diff
@@ -0,0 +1,57 @@
+# unpack the package to ./smarten
+tar -xvf ../.github/workflows/dependencies/EcmaTC49.Smarten.tar
+
+# On some files, smarten can't "round trip", so creates
```
What exactly is meant by "can't round trip" here?

BillWagner replied:
I'd have to dig deeper into the smarten source. On (almost) every file, the smarten_CL writes the same destination file as the source (as we want). One these two, it fails to do that, and writes a different file destination.

Those are the largest files in the standard. I'm suspecting it has something to do with the file size, but I haven't figured that out yet.

BillWagner replied:
Well, after looking at the source, I don't know why those files are generated. These larger files are processed correctly, but an internal diagnostic fails, so the extra file is generated.

I think it can just be removed. I'd like to verify that with @Nigel-Ecma before we merge this.

BillWagner committed:
remove the (RoundTrip) files instead of moving them.

After digging into the code, the RoundTrip files are not the output from the smartening. I'm not sure exactly what they are supposed to represent.

BillWagner committed:
remove untarred exe

Don't add the smarten executable to the pull request content.

Nigel-Ecma commented:
@BillWagner: No it’s not consistent :-( 

The uploaded smarten would roundtrip the then “standard” (so “standard” -> “pass1” -> “pass2” – first two differ, second two are equal).

Of course the contents of “standard” has had a few updates since then, I’ll go dig…

> On 16/03/2022, at 6:09 am, Bill Wagner ***@***.***> wrote:
> 
> 
> @Nigel-Ecma <https://github.com/Nigel-Ecma> One interesting result from testing this locally. Both classes.md and expressions.md wouldn't round trip correctly, so a second file was created instead of overwriting the original. When I looked at the diffs, none of the quotes changed in either file. Is that consistent with your experience? If not, any ideas on how to fix?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/dotnet/csharpstandard/pull/496#issuecomment-1068237772>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABSYVWY5BWVD47FBMNMNN3TVAC74PANCNFSM5QZKUWCQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you were mentioned.

Nigel-Ecma commented:
Helpful information, narrows it down.

The smarten tool does a pre-check of round tripping its input using markdig only, no smartening, and if the output is not exactly the same as the input issues a warning and the extra file so someone (me) can go dig…

> On 16/03/2022, at 7:21 am, Bill Wagner ***@***.***> wrote:
> 
> 
> @BillWagner commented on this pull request.
> 
> In tools/run-smarten.sh <https://github.com/dotnet/csharpstandard/pull/496#discussion_r827286682>:
> 
> > +    "enums"
> +    "delegates"
> +    "exceptions"
> +    "attributes"
> +    "unsafe-code"
> +    "grammar"
> +    "portability-issues"
> +    "standard-library"
> +    "documentation-comments"
> +    "bibliography"
> +    )
> +
> +# unpack the package to ./smarten
> +tar -xvf ../.github/workflows/dependencies/EcmaTC49.Smarten.tar
> +
> +# On some files, smarten can't "round trip", so creates
> Well, after looking at the source, I don't know why those files are generated. These larger files are processed correctly, but an internal diagnostic fails, so the extra file is generated.
> 
> I think it can just be removed. I'd like to verify that with @Nigel-Ecma <https://github.com/Nigel-Ecma> before we merge this.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/dotnet/csharpstandard/pull/496#discussion_r827286682>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABSYVW5NVXNMLBSO62DCKF3VADIKRANCNFSM5QZKUWCQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you were mentioned.

BillWagner commented:
> Helpful information, narrows it down. The smarten tool does a pre-check of round tripping its input using markdig only, no smartening, and if the output is not exactly the same as the input issues a warning and the extra file so someone (me) can go dig…

If it helps, the only diffs between the original and the round trip pass was additional whitespace.

BillWagner committed:
assume all files round trip successfully

BillWagner commented:
Thanks to @Nigel-Ecma 's work in #498, I've removed the extra script logic to move files if the round trip fails.

The script will leave the \*_(RoundTrip).md files in place in the PR if any of the standard source doesn't round-trip correctly. That should help us find any recurrence of issues.

I think this is ready to merge.
